docs/architecture.md

```diff
@@ -0,0 +1,438 @@
+
+- Ops used for modeling PyTorch IValue object graphs (e.g. `torch.nn_module`,
+  `torch.class_type`).
+- `torch.global_slot` and related ops which are used to model an incremental
+  lowering of the IValue object graphs.
+- Ops that are supported in the JIT interpreter directly, and so don't have a
+  corresponding op in the registry (e.g. `torch.prim.If`,
+  `torch.prim.ListConstruct`, `torch.constant.*`)
+- `torch.operator` which is used to represent ops from the registry which
+  haven't been generated by `torch_ods_gen.py`.
+
+### Types
+
+See [TorchTypes.td](https://github.com/llvm/torch-mlir/blob/e322f6a8784009b37aa354abfa9a40a80f30877d/include/torch-mlir/Dialect/Torch/IR/TorchTypes.td#L1)
+
+The `torch` dialect has a complete set of types modeling the PyTorch type
+system, which itself is a strongly typed subset of the Python type system (+
+tensors). These types are almost all 1:1 with the corresponding
+[PyTorch types](https://github.com/pytorch/pytorch/blob/c54d18dbc7bb2f9fdd83c5de529702e5a02295c3/aten/src/ATen/core/jit_type.h#L1).
+
+The one exception where a significant amount of design work has been done in
+Torch-MLIR is the handling of tensors. Torch-MLIR's tensor types allow
+progressive lowering from raw imported IR which maybe be missing shapes, dtypes,
+and value semantics, into the backend contract which provides those. Torch-MLIR
+has two tensor types `ValueTensorType` (`!torch.vtensor`) and
+`NonValueTensorType` (`!torch.tensor`) sharing most of their definition in
+[TorchTypes.td](https://github.com/llvm/torch-mlir/blob/e322f6a8784009b37aa354abfa9a40a80f30877d/include/torch-mlir/Dialect/Torch/IR/TorchTypes.td#L58).
+These types have a common C++ base class
+[`BaseTensorType`](https://github.com/llvm/torch-mlir/blob/e322f6a8784009b37aa354abfa9a40a80f30877d/include/torch-mlir/Dialect/Torch/IR/TorchTypes.h#L40)
+which permits abstracting across them. Both `ValueTensorType` and
+`NonValueTensorType` have an optional list of optional sizes and an optional
+dtype.
+
+## The "backend contract"
+
+See [satisfiesBackendContract](https://github.com/llvm/torch-mlir/blob/114f48e96c578ee76a6f83b3aa4aa229a8d5b76e/lib/Dialect/Torch/Transforms/LowerToBackendContract.cpp#L151)
+
+The backend contract is a normalized form of the `torch` dialect with a set of
+properties that make it easy to lower into various forms such as
+Linalg-on-Tensors, TOSA, MHLO, or other forms that we don't provide out of the
+box. The primary guarantees that we provide Torch-MLIR's backends are:
+
+- All tensors have been converted to value semantics.
+- All tensors have at least a known number of dimensions (i.e. rank), and
+  ideally also have a precise size for each dimension.
+- All tensors have a known dtype.
+- Certain ops have been decomposed to make them easier to handle (this is
+  configurable).
+
+See the extensive comments in the function `satisfiesBackendContract` (and its
+callees) in the `LowerToBackendContract` pass for an extended rationale for
```

Comment:
(at the risk of making the splits even more confusing) Does it make sense to use "middle-end" instead of "backend-contract"? 

Of course this would be a change in many other places than the docs (so not to be considered for this PR), but just wanted to put it out in the context of this architecture doc.

Reply:
I would consider a "middle end" as a place where we do a large number of transformations within a common IR form. Today, the backend contract is just an interface and we don't do much on it. I think if we grow to do more transformations at this level of abstraction, then we can rename it to "middle end".